### PR TITLE
Adapt file methods to new db fim

### LINKED ISSUE #9107: Adapt file methods in FIM DB to the new integrations
|EPIC|Component|
|---|---|
| https://github.com/wazuh/wazuh/issues/9103 | FIM |

### Description

This issue intends to integrate every function and query in `file.cpp` to the new design of FIM DB. It implies:

- Every particular operation over files should be moved to the `DBItemWrapper` class
- Every function used externally should be included in the `FIMDBWrapper` which acts as the external interface.
- Every query should be replaced with equivalent operations performed by `dbsync`
- Every added method should be accompanied by unit tests.

### PR DESCRIPTION
**Pull request by FrancoRivero**
|EPIC|Component|
|---|---|
| https://github.com/wazuh/wazuh/issues/9103 | FIM |

### Description

This pull request adds the following changes:

- [x] Add wrapper between c and c++ in the file.cpp
- [x] Adapt this wrapper to FIM

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Unit tests passed
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [x] cppcheck (static code analysis)

Closes #9107

### CODE REVIEW

File: src/syscheckd/db/src/file.cpp [hunk 1]
```diff
@@ -1,537 +1,254 @@
...
-
-// Convenience macros
-#define fim_db_bind_set_scanned(fim_sql, path) fim_db_bind_path(fim_sql, FIMDB_STMT_SET_SCANNED, path)
-
-#define fim_db_bind_get_path_from_pattern(fim_sql, path) \
-    fim_db_bind_path(fim_sql, FIMDB_STMT_GET_PATH_FROM_PATTERN, path)
-
-// bindings
-
+const auto fileColumnList = R"({"column_list":"[path, mode, last_event, scanned, options, checksum, dev, inode, size,
```
[the remark this answers is not in the document]
Comment (FrancoRivero) [reply src/syscheckd/db/src/file.cpp:16]: This statement is to insert in SQL, "[*]" can be used for the queries but I think that for the insert you have to pass the name of the table variables

File: src/syscheckd/db/include/dbFileItem.hpp [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #define _FILEITEM_HPP
 #include "json.hpp"
 #include "dbItem.hpp"
+#define FIMBD_FILE_TABLE_NAME "file_entry"
```
[the remark this answers is not in the document]
Comment (FrancoRivero) [reply src/syscheckd/db/include/dbFileItem.hpp:16]: Yes, I am changing to fimCommonDefs.h ! Thank you!

File: src/syscheckd/db/include/fimDBHelper.hpp [hunk 3]
```diff
@@ -220,6 +219,19 @@ namespace FIMDBHelper
...
+
+    /**
+    * @brief Create a log message
+    *
+    * @param modules_log_level_t a string with table name
+    * @param msg an array with the column list
+    *
+    */
+    template<typename T>
+    void logErr(const modules_log_level_t logLevel, const std::string& msg)
```
[the remark this answers is not in the document]
Comment (FrancoRivero) [reply src/syscheckd/db/include/fimDBHelper.hpp:231]: You are right ! I am changing it !

File: src/syscheckd/db/include/fimDB.hpp [hunk 4]
```diff
@@ -306,6 +306,14 @@ class FIMDB
             m_stopping = true;
         };
 
+        /**
+         * @brief Its the function to log an error
+         */
+        inline void logErr(const modules_log_level_t logLevel, const std::string& msg)
```
[the remark this answers is not in the document]
Comment (FrancoRivero) [reply src/syscheckd/db/include/fimDB.hpp:312]: You are right ! I am changing it !

File: src/syscheckd/db/src/dbFileItem.cpp [hunk 5]
```diff
@@ -43,26 +43,29 @@ void FileItem::createFimEntry()
...
-    conf.push_back(nlohmann::json::object_t::value_type("uid", m_uid));
-    conf.push_back(nlohmann::json::object_t::value_type("gid", m_gid));
-    conf.push_back(nlohmann::json::object_t::value_type("user_name", m_username));
-    conf.push_back(nlohmann::json::object_t::value_type("group_name", m_groupname));
-    conf.push_back(nlohmann::json::object_t::value_type("hash_md5", m_md5));
-    conf.push_back(nlohmann::json::object_t::value_type("hash_sha1", m_sha1));
-    conf.push_back(nlohmann::json::object_t::value_type("hash_sha256", m_sha256));
-    conf.push_back(nlohmann::json::object_t::value_type("mtime", m_time));
+    conf["table"] = FIMBD_FILE_TABLE_NAME;
+    data.push_back(nlohmann::json::object_t::value_type("path", m_identifier));
```
Comment (Dwordcito) [inline src/syscheckd/db/src/dbFileItem.cpp:50]: what is the expected json for this?

File: src/syscheckd/db/src/file.cpp [hunk 6]
```diff
@@ -1,537 +1,254 @@
...
-        merror("Step error getting count entry data: %s (%d)", sqlite3_errmsg(fim_sql->db), sqlite3_extended_errcode(fim_sql->db));
+        FIMDB::getInstance().logErr(LOG_ERROR, err.what());
     }
 
-    return res;
+    return count;
 }
 
-int fim_db_get_count_file_entry(fdb_t* fim_sql)
+std::vector<std::string> fim_db_get_path_from_pattern(const char* pattern)
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:196]: const char* pattern could be a const std::string& pattern

File: src/syscheckd/db/src/file.cpp [hunk 7]
```diff
@@ -1,537 +1,254 @@
...
 {
-    int res = fim_db_get_count(fim_sql, FIMDB_STMT_GET_COUNT_PATH);
-
-    if (res == FIMDB_ERR)
+    std::vector<std::string> paths;
+    nlohmann::json resultQuery;
+    try
     {
-        merror("Step error getting count entry path: %s (%d)", sqlite3_errmsg(fim_sql->db), sqlite3_extended_errcode(fim_sql->db));
+        auto filter = std::string("WHERE path LIKE") + std::string(pattern);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:202]: const

File: src/syscheckd/db/src/file.cpp [hunk 8]
```diff
@@ -1,537 +1,254 @@
...
+        auto queryFromPattern = FIMDBHelper::dbQuery(FIMBD_FILE_TABLE_NAME, FILE_PRIMARY_KEY, filter, FILE_PRIMARY_KEY);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:203]: const

File: src/syscheckd/db/src/file.cpp [hunk 9]
```diff
@@ -1,537 +1,254 @@
...
+ * @param pattern Pattern that will be used for the LIKE operation.
  *
- * @param fim_sql FIM database structure.
- * @param file_path File name of the file to insert.
- * @param entry FIM entry data structure.
+ * @return a vector with every paths on success, a empty vector otherwise.
  */
-static void fim_db_bind_replace_entry(fdb_t* fim_sql, const char* file_path, const fim_file_data* entry);
-
+std::vector<std::string> fim_db_get_path_from_pattern(const char *pattern);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:26]: Move this out of the __cplusplus, save or avoid it. These are C++ definitions in the implementation of a C interface

File: src/syscheckd/db/src/file.cpp [hunk 10]
```diff
@@ -1,537 +1,254 @@
...
-static int fim_db_set_scanned(fdb_t* fim_sql, const char* path);
-
-int fim_db_get_not_scanned(fdb_t* fim_sql, fim_tmp_file** file, int storage)
+int fim_db_delete_range(const char* pattern,
+                        pthread_mutex_t* mutex,
+                        event_data_t* evt_data,
+                        const directory_t* configuration)
 {
-    if ((*file = fim_db_create_temp_file(storage)) == NULL)
+    auto paths = fim_db_get_path_from_pattern(pattern);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:44]: const and brace initialization.

File: src/syscheckd/db/src/file.cpp (hunk 9 above)
Comment (Dwordcito) [reply src/syscheckd/db/src/file.cpp:26]: use camelCase

File: src/syscheckd/db/src/file.cpp [hunk 11]
```diff
@@ -1,537 +1,254 @@
...
- * @param fim_sql FIM database structure.
- * @param index Index of the particular statement.
- * @param inode Inode of the file.
- * @param dev dev of the file.
+ * @return a vector with paths asociated to the inode.
  */
-static void fim_db_bind_get_inode(fdb_t* fim_sql, int index,
-                                  unsigned long int inode,
-                                  unsigned long int dev);
+std::vector<std::string> fim_db_get_paths_from_inode(unsigned long int inode, unsigned long int dev);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:36]: Ditto

File: src/syscheckd/db/src/file.cpp [hunk 12]
```diff
@@ -1,537 +1,254 @@
...
-int fim_db_process_missing_entry(fdb_t* fim_sql,
-                                 fim_tmp_file* file,
+int fim_db_process_missing_entry(const char* pattern,
                                  pthread_mutex_t* mutex,
-                                 int storage,
                                  event_data_t* evt_data)
 {
-    return fim_db_process_read_file(fim_sql, file, FIM_TYPE_FILE, mutex, fim_delete_file_event, storage, evt_data, NULL,
-                                    NULL);
+    auto paths = fim_db_get_path_from_pattern(pattern);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:66]: const and brace initialization.

File: src/syscheckd/db/src/file.cpp [hunk 13]
```diff
@@ -1,537 +1,254 @@
...
-fim_entry* _fim_db_get_path(fdb_t* fim_sql, const char* file_path)
-{
-    fim_entry* entry = NULL;
-
-    // Clean and bind statements
-    fim_db_clean_stmt(fim_sql, FIMDB_STMT_GET_PATH);
-    fim_db_bind_path(fim_sql, FIMDB_STMT_GET_PATH, file_path);
-
-    if (sqlite3_step(fim_sql->stmt[FIMDB_STMT_GET_PATH]) == SQLITE_ROW)
+    auto paths = fim_db_get_path_from_pattern(pattern);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:86]: const and brace initialization

File: src/syscheckd/db/src/file.cpp [hunk 14]
```diff
@@ -1,537 +1,254 @@
...
+    if (paths.empty())
     {
-        entry = fim_db_decode_full_row(fim_sql->stmt[FIMDB_STMT_GET_PATH]);
+        FIMDB::getInstance().logErr(LOG_ERROR, "No entry found with that pattern");
+        return FIMDB_ERR;
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:90]: avoid multiple returns

File: src/syscheckd/db/src/file.cpp [hunk 15]
```diff
@@ -1,537 +1,254 @@
...
-
-    // Clean statements
-    fim_db_clean_stmt(fim_sql, FIMDB_STMT_GET_PATHS_INODE);
-    fim_db_bind_get_inode(fim_sql, FIMDB_STMT_GET_PATHS_INODE, inode, dev);
-
-    os_calloc(2, sizeof(char*), paths);
-
-    for (i = 0; sqlite3_step(fim_sql->stmt[FIMDB_STMT_GET_PATHS_INODE]) == SQLITE_ROW; i++)
+    nlohmann::json entry_from_path;
+    auto filter = std::string("WHERE path=") + std::string(file_path);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:105]: const and brace initialization.

File: src/syscheckd/db/src/file.cpp [hunk 16]
```diff
@@ -1,537 +1,254 @@
...
+    auto query = FIMDBHelper::dbQuery(FIMBD_FILE_TABLE_NAME, fileColumnList, filter, FILE_PRIMARY_KEY);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:106]: const and brace initialization.

File: src/syscheckd/db/src/file.cpp [hunk 17]
```diff
@@ -1,537 +1,254 @@
...
-        os_realloc(paths, (i + 2) * sizeof(char*), paths);
-
-        p = (char*)sqlite3_column_text(fim_sql->stmt[FIMDB_STMT_GET_PATHS_INODE], 0);
-        sqlite_strdup(p, paths[i]);
+        FIMDBHelper::getDBItem<FIMDB>(entry_from_path, query);
     }
+    catch (DbSync::dbsync_error& err)
+    {
+        FIMDB::getInstance().logErr(LOG_ERROR, err.what());
+        return nullptr;
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:114]: avoid multiple returns.

File: src/syscheckd/db/src/file.cpp [hunk 18]
```diff
@@ -1,537 +1,254 @@
...
-    int nodes_count;
-    int retval = FIMDB_OK;
+    std::vector<std::string> paths;
+    nlohmann::json resultQuery;
 
-    if (syscheck.file_limit_enabled == 0)
+    try
     {
-        return FIMDB_OK;
+        auto filter = std::string("WHERE inode=") + std::to_string(inode) + std::string(" AND dev=") + std::to_string(dev);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:128]: ditto

File: src/syscheckd/db/src/file.cpp [hunk 19]
```diff
@@ -1,537 +1,254 @@
...
+        auto query = FIMDBHelper::dbQuery(FIMBD_FILE_TABLE_NAME, FILE_PRIMARY_KEY, filter, FILE_PRIMARY_KEY);
```
Comment (Dwordcito) [inline src/syscheckd/db/src/file.cpp:129]: ditto

**Review by Dwordcito — CHANGES_REQUESTED** [review]
I see a lot of changes to be made, let's set up a meeting to talk!

File: src/syscheckd/create_db.c [hunk 20]
```diff
@@ -53,8 +53,7 @@ static const char *FIM_EVENT_MODE[] = {
 void update_wildcards_config();
 
 
-void fim_generate_delete_event(fdb_t *fim_sql,
-                               fim_entry *entry,
+void fim_generate_delete_event(char *entry,
```
Comment (antoniomanuelfr) [inline src/syscheckd/create_db.c:56]: Let's use the name `entry` only on `fim_entry` types
```suggestion
void fim_generate_delete_event(char *file_path,
```

**Review by antoniomanuelfr — CHANGES_REQUESTED** [review]
Some minor changes. 

Please, can you run the RTR with the following parameters to follow the code style guide?:
```
python3 build.py --sformat syscheckd/db
```

**Review by JoseAntonioMHerrera — APPROVED** [review]
LGTM!

**Review by jotacarma90 — APPROVED** [review]
LGTM

**Review by jotacarma90 — APPROVED** [review]
Compiation problems solveds, LGTM!

File: src/syscheckd/create_db.c [hunk 21]
```diff
@@ -88,44 +64,33 @@ void fim_generate_delete_event(fdb_t *fim_sql,
...
-            return;
-        }
-        break;
+    // Remove path from the DB.
+    int retVal = fim_db_remove_path(file_path);
 
-    default:
-        break;
+    if (retVal != FIMDB_ERR) {
+        retVal = fim_db_get_path(file_path, callback_data);
```
Comment (JoseAntonioMHerrera) [inline src/syscheckd/create_db.c:88]: I think you need to change the order of those functions calls regarding the extraction and deletion of the file.

Comment (FrancoRivero) [reply src/syscheckd/create_db.c:88]: Yes, this is a bug ! I am fixing that ! Thank you !

**Review by antoniomanuelfr — APPROVED** [review]
LGTM

**Comment by jotacarma90** [comment]
Doing deeper testing with gdb I have found some problems that can occur within the scan. They are due to the constructur of the FileItem class, which receives a fim_entry. Some of the fields of fim_entry could be NULL and this generates a seg fault when trying to initialize the FileItem.

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**